FIX euclidean_distances float32 numerical instabilities

jeremiedbb opened the pull request:
Fixes #9354 
Close #11271 , close #12128

takes only the chunk upcast part of #13410: When input is float32, compute distances on chunks of X and Y upcast to float64.

memory usage of 10% of {X, Y, dist_matrix} with a min of 10Mib.

jeremiedbb committed:
vect vect euclidean

jeremiedbb committed:
vect vect euclidean

Framartin committed:
DOC Fix missing space after backquotes (#13551)

jeremiedbb committed:
minmem

jeremiedbb committed:
docstring

jeremiedbb committed:
lint

jeremiedbb committed:
unrelated

jeremiedbb committed:
update tests + fix int

jeremiedbb committed:
fix chunk size

peay committed:
FIX Explicitly ignore SparseEfficiencyWarning in DBSCAN (#13539)

hermidalc committed:
FIX _estimate_mi discrete_features str and value check (#13497)

* discrete_features str and value check

* Update if logic

* Add discrete_features bad str value test

* Remove unnecessary nested isinstance str check

* Add back nested isinstance str check

* New/updates to tests

* Add v0.21 whats new entry

* Undo v0.21 whats new entry

jnothman commented:
Thank you @jeremiedbb. Some tests are failing. Could you please summarise why you are proposing this solution?

mani2106 committed:
DOC Changed the docstring of class LinearSVR to reflect the default value of epsilon in the code (#13559)

* Update docstring in LinearSVR

Changed the docstring of class LinearSVR to reflect the default value of epsilon in the code.

* Update tree.rst

* Revert "Update tree.rst"

jeremiedbb committed:
fix sparse slice bounds

jeremiedbb commented:
Well, this solution fixes the initial issue, i.e. precision issues in float32 (which was the most requested), while keeping good performance.

for low dimensional data, precision could be increased for float64 at no performance cost, but the cost would be severe for high dimensional data. Not mandatory for 0.21.
performance could be increased (and even be better than current) for float32 on low dimensional data. Not mandatory for 0.21.

jeremiedbb committed:
fix what's new

NicolasHug committed:
DOC Added doc on how to override estimator tags (#13550)

Celelibi commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,123 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    maxmem = max(
+        ((x_density * n_samples_X + y_density * n_samples_Y) * n_features
+         + (x_density * n_samples_X * y_density * n_samples_Y)) / 10,
+        10 * 2**17)
```
Why this choice?
In my tests for #11271 I found that the computation time slowly increase with the memory size. It's the first few plots of this comment: https://github.com/scikit-learn/scikit-learn/pull/11271#issuecomment-398045564.
I'm mostly curious about this choice. I don't think this should prevent this PR from being merged.

jeremiedbb replied:
A fixed 10Mib maxmem can lead to very small chunks when the number of features is very large (extreme case I agree). In that case the drop of performance can be quite severe.

Another possibility could be to have a fixed maxmem with a min chunk_size. I think it's kind of equivalent, but in this case you can have a memory increase bigger than expected whereas in the previous case the memory increase is bounded. That's why I went for the first one.

Celelibi replied:
Large chunks would also decrease cache efficiency. Although I'm not a big fan of having an unbounded amount of additional memory, 10% of the memory already used is probably fair.

When I look at my old benchmarks, I think that there might be something going on under the hood. Maybe the optimal block size is actually constant. Moreover, if the number of features is large, it can also be chunked. That might deserve to be explored as well.
I think we can add to the todo-list to benchmark the influence of the memory size and chunking to get a proper understanding of the performance.

jnothman replied:
If the input data is memmapped, calculating this as a function of n_features may not be fair.

jnothman replied:
I think my comment about memmapping was incorrect in any case.

jnothman replied:
This logic seems to be more complicated than necessary, but okay.

glemaitre replied:
I am fine to merge with this heuristic. 

In the future, I am wondering if we could expose a `chunk_size` parameter which by default do such heuristic and could be fine-tuned by the user. @jeremiedbb mentioned to me that it might not be that easy since it will impact `pairwise_distance` where we would need to give also this parameter, etc. Just a thought.

qinhanmin2014 committed:
MNT Delete _scipy_sparse_lsqr_backport.py (#13569)

kno10 committed:
DOC Clarify eps parameter importance in DBSCAN (#13563)

NicolasHug committed:
DOC Minor typo in GridSearchCV (#13571)

TomDLT committed:
DOC add NicolasHug and thomasjpfan in authors list

Commit:
[MRG] Changed name n_components to n_connected_components in AgglomerativeClustering base class (#13427)

* Changed name n_components to n_connected_components in base class

* Fixed line which exceeded PEP8 max of 79 chars

* Fixed line 818 which exceeded PEP8 max of 79 chars

* Added try and except to provide deprecation warning if  passed

* Updated deprecation and removal version numbers

* Added deprecation of n_components using @property generator

* Makes FeatureAgglomeration class inherit n_connected_components_ attribute from AgglomerativeClustering class

* Added test for DeprecationWarning when trying to access n_components

* Removed @property generator causing linting error

* Fixed typo in test

* Fixed flake8 error due to single line between 2 functions

* Test fix attempt

* Edited test function docstring

* Corrected n_components deprecation test docstring

* Fixed line continuation issue in AgglomerativeClustering base class

* Added deprecation message as part of the @deprecated decorator

* Added  attribute deprecation information in the Attributes section of the AgglomerativeClustering base class docstring

* Added test for deprecation warning message

* Added  attribute deprecation information in the Attributes section of the FeatureAgglomeration base class docstring

* Fixed test issue and added longer match string

* Edited n_components_ deprecation message to add double backticks

* Fixed match string to reflect deprecation message change in test

* Added name to list of contributors

* Documented information in v0.21 changelog

* Added cluster parent folder to documentation in v0.21 changelog

* Removed myself from list of core contributors

* Moved |API| subsection to the end of the list, and changed reference to Github username

* Removed n_components deprecation documentation from FeatureAgglomeration and AgglomerativeClustering class dosctrings

* Fix indentation on _fix_connectivity function call

ClemDoum committed:
FIX non deterministic behavior in SGD (#13422)

rth commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,123 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    n_chunks_X = n_samples_X // chunk_size + (n_samples_X % chunk_size != 0)
+    n_chunks_Y = n_samples_Y // chunk_size + (n_samples_Y % chunk_size != 0)
+
+    for i in range(n_chunks_X):
```
Can we not use `utils.gen_even_slices` here?

jeremiedbb replied:
`gen_batches` since we want a specific batch size, but it's cleaner indeed.

rth commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,123 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    distances = np.empty((n_samples_X, n_samples_Y), dtype=np.float32)
+
+    x_density = X.getnnz() / np.prod(X.shape) if issparse(X) else 1
+    y_density = Y.getnnz() / np.prod(Y.shape) if issparse(Y) else 1
```
Nit: `X.nnz` maybe

wdevazelhes committed:
ENH Allow nd array for CalibratedClassifierCV (#13485)

mani2106 committed:
DOC reference tree structure example from user guide (#13561)

mwestt committed:
DOC correct reference to target in load_linnerud docstring (#13577)

Minor correction to docstring of the function `load_linnerud()` in `base.py`, which was referring to 'targets' instead of 'target'.

adrinjalali committed:
FIX DummyEstimator when y is a 2d column vector (#13545)

* Add column_or_1d to account for dataframe y

* Add test

* make the diff cleaner

* switched pandas import for 2d array

* change test to a simple comparison between 1d and 2d y

* flake8 errors in tests

* Add warning to acieve consistent behaviour as in other classifiers

* Update whats new file

* remove redundant code, fix regression

* remove unnecessary import

* add comment on test

* address comments

* fix what's new entry

* Update doc/whats_new/v0.21.rst

Co-Authored-By: adrinjalali <adrin.jalali@gmail.com>

jeremiedbb committed:
tmp

jeremiedbb committed:
use gen_batches

jeremiedbb committed:
nitpick

samronsin committed:
FIX Take sample weights into account in partial dependence computation for gradient boosting (#13193)

* Replace n_node_samples by weighted_n_node_samples in partial dependence computation

* Add tests for both no-op and real sample weights

* Improve naming and remove useless comment

* Fix small test issues

* Fix test for binary classification

* Add test for regressions based on example from initial issue

* Edit whats_new

* 79

* Simplify test code for regression partial dependence

* PEP8

* Facepalm

* Refer to the public function in whats_new

* Make the sample weight test standalone for further reuse

* Fix PR number

* Testing with L1 relative distance computed as averages

* Testing element-wise

* Fix and simplify unit test for binary classification

* Clarify functional test

amueller committed:
DOC Add lucidfrontier to the emeritus core devs (#13579)

jeremiedbb committed:
FIX MissingIndicator explicit zeros & output shape (#13562)

Commit:
ENH cross_val_predict now handles multi-output predict_proba (#8773)

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+         + (x_density * n_samples_X * y_density * n_samples_Y)) / 10,
+        10 * 2**17)
+
+    # The increase amount of memory is:
```
add "in 8-byte blocks"... in fact where do you account for this?

jeremiedbb replied:
`maxmem` is the max allowed number of 8-byte blocks for extra memory usage.
10Mib is `10 * 2**20` bits = `10 * 2**17` 8-byte blocks.

aditya1702 committed:
Use fixed random seed for generating X in test_mlp.test_gradient() (#13585)

kfrncs committed:
DOC fix typo in comments for svm/classes.py (#13589)

mabdelaal86 committed:
DOC fix typo in contributing.rst (#13593)

okz12 committed:
Improve pipeline parameter error msg (#13536)

DanilBaibak committed:
FEA Parameter for stacking missing indicator into imputer (#12583)

adrinjalali committed:
ENH Convert y in GradientBoosting to float64 instead of float32 (#13524)

Commit:
TST Fixes to make test_pprint.py more resilient to change (#13529)

NicolasHug committed:
FIX Fixed array equality check in pprint (#13584)

Commit:
FEA VotingRegressor (#12513)

adrinjalali committed:
CI update PyPy image to pypy-3-7.0.0 (#13600)

jeremiedbb commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+def _euclidean_distances_upcast_fast(X, XX=None, Y=None, YY=None):
```
it's a residue of my previous PR implementing fast and exact method. I removed it.

jeremiedbb committed:
adress what's comments + clean naming

jeremiedbb commented:
> If the input data is memmapped, calculating this as a function of n_features may not be fair.

Yes that's true. But having a fixed 10Mib can be very bad when the number of features is high because in that case the chunk size is very small.

Do you think 10Mib + a min chunk size would be better (Notice that the 10Mib is not guaranteed in that case) ?

jeremiedbb committed:
Fix empty clusters not correctly relocated when using sample_weight(#13486)

jarrodmillman committed:
DOC Add project_urls to setup.py (#13623)

jeremiedbb committed:
MNT Use a common language_level cython directive (#13630)

qinhanmin2014 committed:
[MRG] DOC Correct coef_ shape in RidgeClassifier (#13633)

* DOC Correct coef_ shape in RidgeClassifier

* address review

thomasjpfan committed:
DOC Adds recent core devs to _contributors.rst (#13640)

kjacks21 committed:
DOC t-SNE perplexity docstring update (#13069)

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -347,9 +347,14 @@ Support for Python 3.4 and below has been officially dropped.
+- |Fix| The function :func:`euclidean_distances`, and therefore and therefore
+  several estimators with ``metric=euclidean``, suffered from numerical
+  precision issues with float32 features. Precision has been increased at the
+  cost of a small drop of performance. :issue:`13554` by :user:`Celelibi` and
```
```suggestion
  cost of a small drop in performance. :issue:`13554` by :user:`Celelibi` and
```

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
             raise ValueError(
                 "Incompatible dimensions for X and X_norm_squared")
+        if XX.dtype == np.float32:
+            XX = None
```
should we raise a warning that these cannot be used??

jeremiedbb replied:
Why not but I'm afraid it could generate a lot of warnings. For instance when pairwise_distances is called in a loop.

One thing we should certainly do is adding an option to `row_norms` to use a float64 accumulator. That's on my todo list :)

jnothman commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+        distances += YY
     np.maximum(distances, 0, out=distances)
 
+    # Ensure that distances between vectors and themselves are set to 0.0.
```
why is this out of the if? It's only applicable if the `if` holds.

(Curiously we currently appear to duplicate this logic in `pairwise_distances_chunked` but not in `pairwise_distances`)

jeremiedbb replied:
> why is this out of the if? It's only applicable if the if holds.

No, even if we use the upcast method, we might end up with negative values due to rounding errors if 2 points are extremely close.

jeremiedbb replied:
you mean zeroing the diagonal ? It applies using both methods.

It's necessary to also do it in pairwise_distances_chunked, because we lose the diagonal info when passing chunks to pairwise_distances.

jnothman replied:
Yes, but I don't think we zero the diagonal in pairwise_distances at all for other metrics

jeremiedbb replied:
I was looking at euclidean_distances, not pairwise_distances... That's strange indeed. Even for metric=euclidean it's not enforced when n_jobs > 1. Let me put that also on my todo list :)

jnothman reviewed (approved):
I think this is looking good enough to merge.

adrinjalali committed:
FIX feature importances in random forests sum up to 1 (#13636)

thomasjpfan commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    # - x_density * batch_size * n_features (copy of chunk of X)
+    # - y_density * batch_size * n_features (copy of chunk of Y)
+    # - batch_size * batch_size (chunk of distance matrix)
+    # Hence x² + (xd+yd)kx = M, where x=batch_size, k=n_features, M=maxmem
```
When a grade school student ask, "Why are we learning the quadratic equation?", we have an answer.

thomasjpfan commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+        if YY.dtype == np.float32:
+            YY = None
+    elif Y.dtype == np.float32:
+        YY = None
```
At first glance, it is not obvious that `None` is used by `_euclidean_distances_upcast`.

jeremiedbb replied:
What do you mean ? `XX` and `YY` are used by `_euclidean_distances_upcast` no matter their value

thomasjpfan replied:
Setting `XX` or `YY` to `None` tells `_euclidean_distances_upcast` to call `row_norms(*_chunk, squared=True)[:, np.newaxis]`. 

Without reading `_euclidean_distances_upcast`, it is difficult to tell why `XX` or `YY` are set to `None`.

glemaitre replied:
I think that we would need a small comment to clarify this part.

thomasjpfan commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -540,41 +540,114 @@ def test_pairwise_distances_chunked():
+                          (np.float64, 1e-8)])
+@pytest.mark.parametrize("dim", [1, 1000000])
+def test_euclidean_distances_extreme_values(dtype, s, dim):
+    # check that euclidean distances is correct where 'fast' method wouldn't be
```
fast method?

jeremiedbb replied:
I made the comment clearer

jeremiedbb committed:
clearer comment

jfbeaumont committed:
DOC Removed a typo from the examples of normalized_mutual_info_score (#13646)

oleksandr-pavlyk committed:
MAINT: n_jobs=-1 replaced with n_jobs=4 in tests (#13644)

This change is to work around the hang

 https://github.com/scikit-learn/scikit-learn/issues/12263

afflicting Windows on machines with > 62 hyperthreads.

DanilBaibak committed:
Add parameter for stacking missing indicator into iterative imputer (#13601)

adrinjalali committed:
FIX ignore single node trees in gbm's feature importances (#13620)

Masstran committed:
DOC typo in sklearn.utils.extmath.weighted_mode (#13655)

1.5+2 = 3.5, not 3.

Commit:
ENH Extension of v_measure_score metric to include beta parameter (#13607)

jeremiedbb committed:
BUG Fix missing 'const' in a few memoryview declaration in trees. (#13626)

jeremiedbb committed:
BLD: check OpenMP support and add a switch to build without it (#13543)

jnothman committed:
FIX initialise Birch centroid_ in all cases (#13651)

Framartin committed:
DOC Improve performance of the plot_rbm_logistic_classification.py example (#13383) (#13648)

praths007 committed:
MNT Import linear assignment from scipy (#13465)

thomasjpfan committed:
DOC Fixes formatting issue in webkit (#13657)

falaktheoptimist committed:
Fixing parameter description (for assume_centered) (#13456)

* Fixing parameter description (for assume_centered)

Fixing parameter description (for assume_centered)

* Update shrunk_covariance_.py

* Update empirical_covariance_.py

rth committed:
MAINT Unvendor joblib (#13531)

kfrncs committed:
Improve comment in setup.py (#13661)

aditya1702 committed:
MAINT Replace absolute imports with relative imports (#13653)

btel committed:
[MRG] Fix various solver issues in ridge_regression and Ridge classes (#13363)

NicolasHug committed:
[MRG + 1] Fix pprint ellipsis (#13436)

orestisfl committed:
DOC Remove space in "cross-entropy" (#13671)

ogrisel committed:
FIX broken references in documentation (#13664)

tommyod committed:
DOC Remove synonyms in documentation of linear models (#13663)

NicolasHug committed:
MNT Correctly handle deprecated attribute warnings and docstrings (#13635)

praths007 committed:
Deprecate "warn_on_dtype" from check_array (#13382)

rebekahkim committed:
Fix MultiOutputClassifier checking for predict_proba method of base estimator (#12222)

#WiMLDS

NicolasHug committed:
fix small latex issue (#13680)

dpwe committed:
Fix sample_weight in label_ranking_average_precision_score (#13447)

amueller committed:
DOC Emeritus core devs final call (#13673)

jnothman commented:
Are we going to get this into 0.21?

thomasjpfan committed:
ENH Add verbose option to Pipeline, FeatureUnion, and ColumnTransformer (#11364)

qinhanmin2014 committed:
MNT Minor clean up in OneVsOneClassifier (#13677)

larsoner committed:
DOC: Fixes for latest numpydoc (#13670)

jarrodmillman committed:
Typo (#13693)

adrinjalali committed:
FIX make sure vectorizers read data from file before analyzing (#13641)

bharatr21 committed:
DOC: Add SLEP and Governance in Dev Docs (#13688)

adrinjalali committed:
MAINT: minor fix to whats_new (#13695)

Thanks

jnothman committed:
DOC Describe what's new categories (#13697)

qinhanmin2014 committed:
MNT Minor clean up in OneVsRestClassifier (#13675)

jnothman commented:
Are you able to look at this, @qinhanmin2014?

qinhanmin2014 commented:
> Are you able to look at this, @qinhanmin2014?

I won't be available until this weekend (will try to merge OPTICS these days).

xinyuliu12 committed:
MAINT: reduce example execution time of plot_image_denoising.py (#13683)

jnothman committed:
DOC better wording in changelog legend

eamanu committed:
ENH Support Haversine distance in pairwise_distances (#12568)

NicolasHug committed:
FEA Partial dependence plots (#12599)

jnothman commented:
@rth are you okay with this?

thomasjpfan committed:
MAINT Uses debian stretch for circleci doc building (#13642)

NicolasHug committed:
FEA Add a stratify option to utils.resample (#13549)

rth commented:
> Are we going to get this into 0.21?

It would good I think.

Will review this in the next couple of days.

glemaitre commented on doc/whats_new/v0.21.rst:
```diff
@@ -347,9 +347,14 @@ Support for Python 3.4 and below has been officially dropped.
+  in version 0.21 and will be removed in version 0.23. :issue:`10580` by
+  :user:`Reshama Shaikh <reshamas>` and :user:`Sandra Mitrovic <SandraMNE>`.
+
+- |Fix| The function :func:`euclidean_distances`, and therefore and therefore
```
```suggestion
- |Fix| The function :func:`euclidean_distances`, and therefore
```

glemaitre commented on doc/whats_new/v0.21.rst:
```diff
@@ -347,9 +347,14 @@ Support for Python 3.4 and below has been officially dropped.
+  :user:`Reshama Shaikh <reshamas>` and :user:`Sandra Mitrovic <SandraMNE>`.
+
+- |Fix| The function :func:`euclidean_distances`, and therefore and therefore
+  several estimators with ``metric=euclidean``, suffered from numerical
```
```suggestion
  several estimators with ``metric='euclidean'``, suffered from numerical
```

glemaitre commented on doc/whats_new/v0.21.rst:
```diff
@@ -347,9 +347,14 @@ Support for Python 3.4 and below has been officially dropped.
+
+- |Fix| The function :func:`euclidean_distances`, and therefore and therefore
+  several estimators with ``metric=euclidean``, suffered from numerical
+  precision issues with float32 features. Precision has been increased at the
```
```suggestion
  precision issues with ``float32`` features. Precision has been increased at the
```

glemaitre commented on sklearn/metrics/pairwise.py:
```diff
@@ -203,7 +203,7 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     Returns
     -------
-    distances : {array, sparse matrix}, shape (n_samples_1, n_samples_2)
+    distances : array, shape (n_samples_1, n_samples_2)
```
I put the comment here but I think that it should be written above.

I would either add a note (using sphinx syntax) or within the parameter description that `X_norm_squared` and `Y_norm_squared` will be ignored if they are passed as `float32` since they will lead to inaccurate distances.

glemaitre commented on sklearn/metrics/pairwise.py:
```diff
@@ -231,34 +231,115 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    Assumes XX and YY have float64 dtype or are None.
+
+    X and Y are upcast to float64 by chunks, which size is chosen to limit
+    memory increase by approximately 10%.
```
```suggestion
    memory increase by approximately 10% (bounded to 10Mb).
```

glemaitre commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -540,41 +540,114 @@ def test_pairwise_distances_chunked():
+    Y = y_array_constr(Y)
+    distances = euclidean_distances(X, Y)
+
+    assert_allclose(distances, expected, rtol=1e-6)
```
a small comments regarding `rtol=1e-6`

glemaitre commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -540,41 +540,114 @@ def test_pairwise_distances_chunked():
+    assert distances.dtype == dtype
+
+
+@pytest.mark.parametrize("dtype, s",
```
Maybe we could have a better name for `s`

glemaitre commented on sklearn/metrics/tests/test_pairwise.py:
```diff
@@ -540,41 +540,114 @@ def test_pairwise_distances_chunked():
+
+@pytest.mark.parametrize("dtype, s",
+                         [(np.float32, 1e-4),
+                          (np.float64, 1e-8)])
```
```suggestion
                          pytest.param(np.float64, 1e-8, marks=pytest.mark.xfail('failing due to lack of precision'))])
```

glemaitre reviewed (changes requested):
LGTM. Only nitpicks.

glemaitre committed:
Update doc/whats_new/v0.21.rst

Co-Authored-By: jeremiedbb <34657725+jeremiedbb@users.noreply.github.com>

glemaitre committed:
Update doc/whats_new/v0.21.rst

Co-Authored-By: jeremiedbb <34657725+jeremiedbb@users.noreply.github.com>

glemaitre committed:
Update doc/whats_new/v0.21.rst

Co-Authored-By: jeremiedbb <34657725+jeremiedbb@users.noreply.github.com>

rajdeepd committed:
additional tests for mean_shift algo (#13179)

jnothman committed:
DOC use :pr: rather than :issue: in what's new (#13701)

glemaitre committed:
DOC fix remaining :issue: (#13716)

glemaitre commented on doc/whats_new/v0.21.rst:
```diff
@@ -347,9 +347,14 @@ Support for Python 3.4 and below has been officially dropped.
+- |Fix| The function :func:`euclidean_distances`, and therefore
+  several estimators with ``metric='euclidean'``, suffered from numerical
+  precision issues with ``float32`` features. Precision has been increased at the
+  cost of a small drop of performance. :issue:`13554` by :user:`Celelibi` and
```
```suggestion
  cost of a small drop of performance. :pr:`13554` by :user:`Celelibi` and
```

jeremiedbb committed:
merge

jeremiedbb committed:
adress comments

jeremiedbb committed:
Merge remote-tracking branch 'upstream/master' into euclidean_dist_upcast

jeremiedbb commented:
I messed up :/  but the diff is correct

jnothman commented:
We should take that review as Approve, @glemaitre?

rth reviewed (commented):
Thanks to work working on this @jeremiedbb, and for all the insightful comments/discussions by reviewers.

Overall looks good to me. Below are a few comments mostly about readability..

glemaitre commented:
Remaining the comments of @rth

jnothman committed:
astype=False per Roman's comment

jnothman reviewed (commented):
I made the astype(copy=False) change under the assumption that this is the only substantive change blocking merge.

rth commented:
> I made the astype(copy=False) change under the assumption that this is the only substantive change blocking merge.

Thanks, feel free to merge if this PR is blocking the release.

jnothman commented:
I consider #13739 to be the blocker.

jeremiedbb commented on sklearn/metrics/pairwise.py:
```diff
@@ -224,41 +231,125 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+                d += XX_chunk
+                d += YY_chunk
+
+            distances[x_slice, y_slice] = d.astype(np.float32)
```
this function only computes euclidean distances on float64 and downcast it to float32 at the end so actually there will always be a copy. Actually I find it more clear to not add the `copy=False` parameter to emphasize that

jeremiedbb commented on sklearn/metrics/pairwise.py:
```diff
@@ -224,41 +231,125 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+            if X is Y and j < i:
+                # when X is Y the distance matrix is symmetric so we only need
+                # to compute half of it.
+                d = distances[y_slice, x_slice].T
```
> still not confident that when we compute distances[x_slice, y_slice], distances[y_slice, x_slice] is already computed

It is since we compute the chunks with `i` (rows) as the outer loop. It implies that all chunks of the upper right triangular part of the distance matrix are computed before their symmetric chunk in the lower left triangular part.

jeremiedbb commented on sklearn/metrics/pairwise.py:
```diff
@@ -224,41 +231,125 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    # 10Mib)
+    maxmem = max(
+        ((x_density * n_samples_X + y_density * n_samples_Y) * n_features
+         + (x_density * n_samples_X * y_density * n_samples_Y)) / 10,
```
> This makes the assumption that sizeof(dtype) == 8 without making it appear in the computation, I think?

This function is only used for float32 X and Y. I've explained it in its docstring.

> for sparse it's a roughly 10% if it is CSR and we account for X.data, X.indptr in 32 bit, but again it's far from obvious for a casual reader.

When you change the type of the csr matrix, only a copy of `data` is made. `indices` and `indptr` are still int, so I think this formula is correct

rth replied:
> When you change the type of the csr matrix, only a copy of data is made.

Are you sure?

```py
>>> import scipy.sparse
>>> import numpy as nop
>>> import numpy as np
>>> X = scipy.sparse.csr_matrix(np.ones((10, 10)))
>>> X.dtype
dtype('float64')
>>> Y = X.astype(np.float32)
>>> np.shares_memory(Y.data, Y.data)
True
>>> np.shares_memory(Y.data, X.data)
False
>>> np.shares_memory(Y.indices, X.indices)
False
```

That's not critical, my point here is that it might be beter (not necessarily in this PR) to use `ndarray.nbytes` than trying to re-compute that from scratch. If we get this calculation wrong we won't likely know, since no test will fail and we will only reason in wrong sizes of memory cache.

jeremiedbb commented on sklearn/metrics/pairwise.py:
```diff
@@ -224,41 +231,125 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
+    #                                 xd=x_density and yd=y_density
+    tmp = (x_density + y_density) * n_features
+    batch_size = (-tmp + np.sqrt(tmp**2 + 4 * maxmem)) / 2
+    batch_size = max(int(batch_size), 1)
```
> We don't really care about getting maxmem exactly right, do we? If it's about cache friendliness the inner loop would matter most I think?

Not exactly but we want to upper bound it I guess. With your formulation, we don't control it, since it does not take account of the memory of the computed chunk of the distance matrix of size `batch_size_x * batch_size_y`

rth replied:
Fair enough.

jeremiedbb commented:
I'm ok letting astype(copy=False), but I want to remind that this function is only used on float32 data so the copy will always be made, and I find more confusing since it may give the impression that it's possible to have no copy.

jnothman commented:
So then remove the copy=True. Thanks. Happy to merge when you think you
have addressed Roman's comments.

jeremiedbb committed:
Mib -> MiB

jeremiedbb commented:
I think I have. @rth do you agree with my answers to your comments ?

rth commented:
> but I want to remind that this function is only used on float32 data so the copy will always be made, 

In general, a copy will be made, but when `X is Y`, on the upper triangular matrix, this used to do,
```
distances[x_slice, y_slice] = distances[y_slice, x_slice].astype(np.float32)
```
which does make 1 extra unnecessary copy since `distances` is already float32.

> and I find more confusing since it may give the impression that it's possible to have no copy.

Yes, hence my comments about splitting that part of code a bit more. But it's not critical.

jeremiedbb committed:
back to copy=False

jeremiedbb commented:
you're right. I put the `copy=False` back :)

rth commented on sklearn/metrics/pairwise.py:
```diff
@@ -302,7 +302,7 @@ def _euclidean_distances_upcast(X, XX=None, Y=None, YY=None):
 
     # Allow 10% more memory than X, Y and the distance matrix take (at least
-    # 10Mib)
+    # 10MiB)
```
Well, 10MiB is 80Mib. Wouldn't you have to change the calculations below as well? :)

jeremiedbb replied:
I don't think so, I wrote Mib but it was actually MiB. MiB is 1024 * 1024 bytes, right ? If so, that's what I used

rth reviewed (approved):
Last comment apart from that LGTM.

rth commented:
Merging. Thanks @jeremiedbb and @Celelibi !
